### PR TITLE
feat(cp): add basic connection pool

### PR DESCRIPTION
**Pull request by RomanAbaev**
i have created basic caonnection pool
q: i don't really know how can i test this connection pool, would be great if you hint me, thanks

### CODE REVIEW

File: src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    private static final int CAPACITY = 10;
+
+    private ConcurrentLinkedQueue<Connection> availableConnections;
+    private Vector<Connection> usedConnections;
+    private String url = null;
+    private String user = null;
+    private String password = null;
+
+    private static volatile ConnectionPool instance;
```
Comment (akalji) [inline src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java]: Do not use volatile with Object =)

File: src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Properties;
+import java.util.Vector;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class ConnectionPool {
+
+    private static final int CAPACITY = 10;
+
+    private ConcurrentLinkedQueue<Connection> availableConnections;
+    private Vector<Connection> usedConnections;
```
Comment (akalji) [inline src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java]: Using interfaces is recommended. Also Vector class is depricated

**Review by dodo0101 — APPROVED** [review]

**Review by Dji-key — CHANGES_REQUESTED** [review]
No

**Review by dodo0101 — CHANGES_REQUESTED** [review]
please change

**Comment by RomanAbaev** [comment]
> please change

okay, chief!

**Review by dodo0101 — APPROVED** [review]

File: src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+public class ConnectionPool implements AutoCloseable {
```
Comment (Dji-key) [inline src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java]: No

**Review by Dji-key — CHANGES_REQUESTED** [review]
Again

File: src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    private static String url = null;
+    private static String user = null;
+    private static String password = null;
+
+    private static ConnectionPool instance;
+
+    public static ConnectionPool getInstance() {
+
+        ConnectionPool localInstance = instance;
+        if (localInstance == null) {
```
Comment (Dji-key) [inline src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java]: Lazy initialization is deprecated by us

File: src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            }
+
+
+            localInstance = instance;
+            if (localInstance == null) {
+                instance = localInstance = new ConnectionPool();
+            }
+        }
+        return localInstance;
+    }
```
Comment (Dji-key) [inline src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java:56]: Change lazy initialization to eager

File: src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                property.load(fis);
+
+                url = property.getProperty("db.url");
+                user = property.getProperty("db.user");
+                password = property.getProperty("db.password");
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+
+            availableConnections = new ArrayList<>();
```
Comment (Dji-key) [inline src/main/java/com/epam/lab/library/connectionpool/ConnectionPool.java:40]: Change to LinkedList cause remove too slow vs LinkedList.addLast / removeLast

**Review by Dji-key — APPROVED** [review]

**Pull request merged**